src/CustomElementInternals.js

```diff
@@ -0,0 +1,294 @@
+import CustomElementDefinition from './CustomElementDefinition';
+import * as Utilities from './Utilities';
+import * as CESymbols from './CustomElementInternalSymbols';
+import CEState from './CustomElementState';
+
+export default class CustomElementInternals {
+  constructor() {
+    /** @type {!Map<string, !CustomElementDefinition>} */
+    this._localNameToDefinition = new Map();
+
+    /** @type {!Map<!Function, !CustomElementDefinition>} */
+    this._constructorToDefinition = new Map();
+
+    /** @type {!Array<!function(!Node)>} */
+    this._patches = [];
+  }
+
+  /**
+   * @param {string} localName
+   * @param {!CustomElementDefinition} definition
+   */
+  setDefinition(localName, definition) {
+    this._localNameToDefinition.set(localName, definition);
+    this._constructorToDefinition.set(definition.constructor, definition);
+  }
+
+  /**
+   * @param {string} localName
+   * @return {!CustomElementDefinition|undefined}
+   */
+  localNameToDefinition(localName) {
+    return this._localNameToDefinition.get(localName);
+  }
+
+  /**
+   * @param {!Function} constructor
+   * @return {!CustomElementDefinition|undefined}
+   */
+  constructorToDefinition(constructor) {
+    return this._constructorToDefinition.get(constructor);
+  }
+
+  /**
+   * @param {!function(!Node)} listener
+   */
+  addPatch(listener) {
+    this._patches.push(listener);
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  patchTree(node) {
+    Utilities.walkDeepDescendants(node, node => this.patch(node));
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  patch(node) {
+    if (node[CESymbols.patched]) return;
```

Comment:
Probably best perf if this can immediately bail on failing a boolean check to a flag that is set if instance patching is needed. I think having a set of patches probably suffers from YAGNI.